### PR TITLE
Exitus `UnicastBus`

### PR DESCRIPTION
**Pull request by danielmarbach**
You served us well dear `UnicastBus` but now it is time to be banned away from public. In your honor I brought here a little poem.

**A Happy Man By Edwin Arlington Robinson**
> When these graven lines you see, 
> Traveller, do not pity me; 
> Though I be among the dead, 
> Let no mournful word be said.

> Children that I leave behind, 
> And their children, all were kind; 
> Near to them and to my wife, 
> I was happy all my life.

> My three sons I married right, 
> And their sons I rocked at night; 
> Death nor sorrow never brought 
> Cause for one unhappy thought.

> Now, and with no need of tears, 
> Here they leave me, full of years,-- 
> Leave me to my quiet rest 
> In the region of the blest.
>
> http://www.familyfriendpoems.com/poem/a-happy-man-by-edwin-arlington-robinson

The purpose of this PR is to remove the public facing area of the library and finally get rid of UnicastBus. It has been abused too long. With the pipeline, extension objects and features we should have enough hooks to implement everything done on UnicastBus. Fare well my friend!

Started as part of https://github.com/Particular/NServiceBus/commit/ff3da757b3f024128c62759ca80a70d8e32d9103#commitcomment-12311161

@Particular/nservicebus please review

### CODE REVIEW

File: src/NServiceBus.Core/Pipeline/BehaviorContextStacker.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@
     /// <summary>
     /// A stack of <see cref="BehaviorContext"/>s.
     /// </summary>
-    public class BehaviorContextStacker : IDisposable
+    internal class BehaviorContextStacker : IDisposable
```
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Pipeline/BehaviorContextStacker.cs]: Nitpicking: `internal` is the default

Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/BehaviorContextStacker.cs]: My old style cop obsession still gets me from time to time :(

Am Donnerstag, 23. Juli 2015 schrieb Szymon Pobiega :

File: src/NServiceBus.Core/Unicast/Config/UnicastBus.cs [hunk 2]
```diff
@@ -94,9 +92,9 @@ protected internal override void Setup(FeatureConfigurationContext context)
             }
         }
 
-        Unicast.UnicastBus CreateBus(IBuilder builder)
+        static UnicastBusInternal CreateBus(IBuilder builder)
```
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Unicast/Config/UnicastBus.cs]: Can we now remove this along with the registrations above? If we just `ConfigureComponent<UnicastBusInternal>()` it should just work.

**Comment by SzymonPobiega** [comment]
Yay!

**Comment by jbogard** [comment]
Wait, so does this mean that there's never going to be a MulticastBus?!?!

File: src/NServiceBus.Core/Pipeline/BehaviorContextStacker.cs (hunk 1 above)
Comment (adamralph) [inline src/NServiceBus.Core/Pipeline/BehaviorContextStacker.cs]: I realise it's not the standard in the Particular org, but I'm :+1: for explicit access modifiers always.

**Comment by andreasohlund** [comment]
The only thing we need to figure out is if/how we want to provide access to the "builder"

Today we have the `var builder = ((UnicastBus)bus).Builder` hack.

Options

1. Tell users to use a explicit container if they need access to the container outside their handlers
2. Provice some api to access it

I vote for no.1 which would make it just a doco issue

File: src/NServiceBus.Core/Pipeline/BehaviorContextStacker.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/BehaviorContextStacker.cs]: I've had that argument without success

Am Freitag, 24. Juli 2015 schrieb Adam Ralph :

**Comment by SeanFeldman** [comment]
Should `ReplacementTypeOrMember` of `[ObsoleteEx(...)]` indicate what the replacement should be and if no such, be empty instead of repeating API method/property that is being obsoleted?

**Comment by danielmarbach** [comment]
@SeanFeldman which one do you mean? I tried to point to `IBus` members which are still available

**Comment by danielmarbach** [comment]
@andreasohlund Not sure I agree. It was a hack. Did we actually document it? I couldn't find it here:

http://docs.particular.net/nservicebus/upgrades/4to5

So if not documented. It doesn't exists ;)

**Comment by SzymonPobiega** [comment]
@danielmarbach I think we need to mention that hack. Although it was never official, I remember it being mentioned a few times on the google group by Particular people.

**Comment by danielmarbach** [comment]
Ok. I'll open up a PR to the upgrade guide

**Comment by danielmarbach** [comment]
@SzymonPobiega @andreasohlund I opened up a [PR for the doco](https://github.com/Particular/docs.particular.net/pull/716). As soon as this one is merged we can go ahead with the doco merge/ @SzymonPobiega can you review again and merge?

**Comment by SzymonPobiega** [comment]
Amen!

**Pull request merged**